src/mesh/mesh.cpp

```diff
@@ -901,11 +908,34 @@ void Mesh::EnrollBndryFncts_(ApplicationInput *app_in) {
 void Mesh::ApplyUserWorkBeforeOutput(Mesh *mesh, ParameterInput *pin,
                                      SimTime const &time) {
   // call Mesh version
-  mesh->UserMeshWorkBeforeOutput(mesh, pin, time);
+  if (mesh->UserMeshWorkBeforeOutput) {
+    mesh->UserMeshWorkBeforeOutput(mesh, pin, time);
+  }
 
   // call MeshBlock version
   for (auto &pmb : block_list) {
-    pmb->UserWorkBeforeOutput(pmb.get(), pin);
+    if (pmb->UserWorkBeforeOutput) {
+      pmb->UserWorkBeforeOutput(pmb.get(), pin, time);
+    }
+  }
+}
+
+//----------------------------------------------------------------------------------------
+// \!fn void Mesh::ApplyUserWorkBeforeRestartOutput
+// \brief Apply Mesh and Meshblock versions of UserWorkBeforeRestartOutput
+void Mesh::ApplyUserWorkBeforeRestartOutput(Mesh *mesh, ParameterInput *pin,
+                                            SimTime const &time,
+                                            OutputParameters *pparams) {
+  // call Mesh version
+  if (mesh->UserMeshWorkBeforeRestartOutput) {
+    mesh->UserMeshWorkBeforeRestartOutput(mesh, pin, time, pparams);
+  }
+
+  // call MeshBlock version
+  for (auto &pmb : block_list) {
+    if (pmb->UserWorkBeforeRestartOutput) {
```

Comment:
Do we want to allow/support both callbacks to be defined?
In many other places, we restricted the use to one version so that's clear what's going to happen.

Reply:
Yeah let me remove the `MeshBlock` version of `UserWorkBeforeRestartOutput` I think that is purely redundant. But I'm hesitant to remove the existing `MeshBlock` version of `UserWorkBeforeOutput`-- if we are going to refactor existing hooks I think that should be a separate PR.